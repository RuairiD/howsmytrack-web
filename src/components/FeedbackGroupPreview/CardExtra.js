import React from "react";
import dateFormat from "dateformat";

import { Button, Typography } from "antd";
import { Div } from "lemon-reset";

const formatTimeCreated = (timeCreated) => dateFormat(
    new Date(Date.parse(timeCreated)),
    "mmmm dS yyyy",
);

const CardExtra = ({ timeCreated }) => (
    <Div style={{ display: "flex", marginTop: "1em" }}>
        <Typography.Text style={{
            marginRight: "auto",
            marginTop: "auto",
            marginBottom: "auto",
        }}
        >
            {formatTimeCreated(timeCreated)}
        </Typography.Text>
        <Typography.Text style={{
            marginLeft: "auto",
            marginTop: "auto",
            marginBottom: "auto",
        }}
        >
            <Button type="link">View Group</Button>
        </Typography.Text>
    </Div>
);

export default CardExtra;
