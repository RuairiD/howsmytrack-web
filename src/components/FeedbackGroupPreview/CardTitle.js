import React from "react";

import { Badge, Typography } from "antd";
import { A, Div } from "lemon-reset";

const CardTitle = ({ feedbackGroupId, name, unreadReplies }) => (
    <A href={`/group/${feedbackGroupId}`}>
        <Div style={{ display: "flex" }}>
            <Typography.Title level={4} style={{ marginRight: "auto", marginBottom: "auto", marginTop: "auto" }}>
                {name}
            </Typography.Title>
            <Badge
                count={unreadReplies}
                style={{ marginLeft: "auto", marginBottom: "auto", marginTop: "auto" }}
            />
        </Div>
    </A>
);

export default CardTitle;
