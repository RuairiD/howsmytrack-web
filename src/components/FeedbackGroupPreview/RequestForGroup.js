import React from "react";

import { Row, Col, Typography } from "antd";

import MediaEmbed from "../MediaEmbed/MediaEmbed";

const RequestForGroup = ({ mediaUrl, mediaType }) => {
    if (mediaUrl) {
        return (
            <Row gutter={[16, 16]}>
                <Col>
                    <Typography.Text strong>You submitted:</Typography.Text>
                    <MediaEmbed
                        mediaUrl={mediaUrl}
                        mediaType={mediaType}
                        size="small"
                    />
                </Col>
            </Row>
        );
    }
    return (
        <Row gutter={[16, 16]}>
            <Col>
                <Typography.Text strong>You did not submit a track for this group.</Typography.Text>
            </Col>
        </Row>
    );
};

export default RequestForGroup;
