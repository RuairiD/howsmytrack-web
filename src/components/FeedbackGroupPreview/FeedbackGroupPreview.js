import React from "react";

import { Button, Card, Row, Col, Icon, Progress, Tooltip, Typography } from "antd";
import { A, Div } from "lemon-reset";

import type { FeedbackRequestSummaryProps } from "../FeedbackRequestSummary/FeedbackRequestSummary";

import CardTitle from "./CardTitle";
import CardExtra from "./CardExtra";
import RequestForGroup from "./RequestForGroup";

const FEEDBACK_FOR_THEM_TOOLTIP = "How many tracks you've written feedback for in this group.";
const FEEDBACK_FOR_YOU_TOOLTIP = "How many members of this group have written feedback for your track (this number is hidden if you have not yet written feedback for everyone in your group)";

export type FeedbackGroupPreviewProps = {
    feedbackGroupId: number,
    name: string,
    timeCreated: string,
    // The user's original request for the group
    feedbackRequestSummary: FeedbackRequestSummaryProps,
    // Number of users in the group with tracks
    userCount: number,
    // Number of users in the group without tracks
    tracklessUserCount: number,
    // Number of submissions in the group for which the logged-in user has submitted feedback
    userFeedbackCount: number,
    // Number of group members who have submitted feedback for the user
    feedbackResponseCount: number,
    // Number of replies sent to the user that haven't been read yet.
    unreadReplies: number,
};

/*
 * Component for inline group preview shown on FeedbackGroupsPage
 */
const FeedbackGroupPreview = ({
    feedbackGroupId,
    name,
    timeCreated,
    feedbackRequestSummary,
    userCount,
    tracklessUserCount,
    userFeedbackCount,
    feedbackResponseCount,
    unreadReplies,
}: FeedbackGroupPreviewProps) => (
    <Div>
        <Card title={(
            <CardTitle
                feedbackGroupId={feedbackGroupId}
                name={name}
                unreadReplies={unreadReplies}
            />
        )}
        >
            <A href={`/group/${feedbackGroupId}`}>
                <RequestForGroup mediaUrl={feedbackRequestSummary.mediaUrl} mediaType={feedbackRequestSummary.mediaType} />
                <Row gutter={[16, 16]}>
                    <Col>
                        <Typography.Text strong>Feedback for them</Typography.Text>
                        <Tooltip
                            title={FEEDBACK_FOR_THEM_TOOLTIP}
                        >
                            <Button type="link"><Icon type="question-circle" /></Button>
                        </Tooltip>
                        <Progress
                            percent={100 * (userFeedbackCount / userCount)}
                            format={() => `${userFeedbackCount}/${userCount}`}
                            width={80}
                        />
                    </Col>
                </Row>
                {feedbackRequestSummary.mediaUrl && (
                    <Row gutter={[16, 16]}>
                        <Col>
                            <Typography.Text strong>Feedback for you</Typography.Text>
                            <Tooltip
                                title={FEEDBACK_FOR_YOU_TOOLTIP}
                            >
                                <Button type="link"><Icon type="question-circle" /></Button>
                            </Tooltip>
                            <Progress
                                percent={100 * (feedbackResponseCount / (userCount + tracklessUserCount))}
                                format={() => `${feedbackResponseCount}/${userCount + tracklessUserCount}`}
                                width={80}
                            />
                        </Col>
                    </Row>
                )}
                <Card.Meta description={<CardExtra timeCreated={timeCreated} />} />
            </A>
        </Card>
    </Div>
);

export default FeedbackGroupPreview;
