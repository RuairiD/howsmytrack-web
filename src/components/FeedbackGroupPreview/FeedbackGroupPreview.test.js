import React from "react";
import { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import FeedbackGroupPreview from "./FeedbackGroupPreview";

describe("FeedbackGroupPreview", () => {
    it("displays a feedback group's preview, including a media preview", () => {
        const wrapper = shallow(
            <FeedbackGroupPreview
                feedbackGroupId={1901}
                name="name"
                timeCreated={123456}
                feedbackRequestSummary={{
                    feedbackRequestId: 808,
                    mediaUrl: "mediaUrl",
                    mediaType: "mediaType",
                    feedbackPrompt: "feedbackPrompt",
                    emailWhenGrouped: false,
                    genre: "genre",
                }}
                userCount={4}
                tracklessUserCount={1}
                userFeedbackCount={3}
                feedbackResponseCount={3}
                unreadReplies={2}
            />,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("displays a feedback group's preview for a trackless request without media", () => {
        const wrapper = shallow(
            <FeedbackGroupPreview
                feedbackGroupId={1901}
                name="name"
                timeCreated={123456}
                feedbackRequestSummary={{
                    feedbackRequestId: 808,
                    mediaUrl: null,
                    mediaType: null,
                    feedbackPrompt: null,
                    emailWhenGrouped: false,
                    genre: "genre",
                }}
                userCount={4}
                tracklessUserCount={null}
                userFeedbackCount={3}
                feedbackResponseCount={3}
                unreadReplies={2}
            />,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("formats counts for Progress meters", () => {
        const wrapper = mount(
            <FeedbackGroupPreview
                feedbackGroupId={1901}
                name="name"
                timeCreated={123456}
                feedbackRequestSummary={{
                    feedbackRequestId: 808,
                    mediaUrl: "mediaUrl",
                    mediaType: "mediaType",
                    feedbackPrompt: "feedbackPrompt",
                    emailWhenGrouped: false,
                    genre: "genre",
                }}
                userCount={4}
                tracklessUserCount={1}
                userFeedbackCount={3}
                feedbackResponseCount={2}
                unreadReplies={2}
            />,
        );

        expect(wrapper.find("Progress").length).toBe(2);
        expect(wrapper.find("Progress").get(0).props.percent).toBe(75);
        expect(wrapper.find("Progress").get(0).props.format()).toBe("3/4");
        expect(wrapper.find("Progress").get(1).props.percent).toBe(40);
        expect(wrapper.find("Progress").get(1).props.format()).toBe("2/5");
    });
});
