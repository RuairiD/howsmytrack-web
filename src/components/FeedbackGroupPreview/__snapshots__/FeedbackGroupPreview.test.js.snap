// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedbackGroupPreview displays a feedback group's preview for a trackless request without media 1`] = `
<Div
  className=""
>
  <Card
    title={
      <CardTitle
        feedbackGroupId={1901}
        name="name"
        unreadReplies={2}
      />
    }
  >
    <A
      className=""
      href="/group/1901"
    >
      <RequestForGroup
        mediaType={null}
        mediaUrl={null}
      />
      <Row
        gutter={
          Array [
            16,
            16,
          ]
        }
      >
        <Col>
          <Text
            strong={true}
          >
            Feedback for them
          </Text>
          <Tooltip
            arrowPointAtCenter={false}
            autoAdjustOverflow={true}
            mouseEnterDelay={0.1}
            mouseLeaveDelay={0.1}
            placement="top"
            title="How many tracks you've written feedback for in this group."
            transitionName="zoom-big-fast"
          >
            <Button
              block={false}
              ghost={false}
              htmlType="button"
              loading={false}
              type="link"
            >
              <Icon
                type="question-circle"
              />
            </Button>
          </Tooltip>
          <Progress
            format={[Function]}
            gapDegree={0}
            percent={75}
            showInfo={true}
            size="default"
            strokeLinecap="round"
            trailColor="#f3f3f3"
            type="line"
            width={80}
          />
        </Col>
      </Row>
      <Meta
        description={
          <CardExtra
            timeCreated={123456}
          />
        }
      />
    </A>
  </Card>
</Div>
`;

exports[`FeedbackGroupPreview displays a feedback group's preview, including a media preview 1`] = `
<Div
  className=""
>
  <Card
    title={
      <CardTitle
        feedbackGroupId={1901}
        name="name"
        unreadReplies={2}
      />
    }
  >
    <A
      className=""
      href="/group/1901"
    >
      <RequestForGroup
        mediaType="mediaType"
        mediaUrl="mediaUrl"
      />
      <Row
        gutter={
          Array [
            16,
            16,
          ]
        }
      >
        <Col>
          <Text
            strong={true}
          >
            Feedback for them
          </Text>
          <Tooltip
            arrowPointAtCenter={false}
            autoAdjustOverflow={true}
            mouseEnterDelay={0.1}
            mouseLeaveDelay={0.1}
            placement="top"
            title="How many tracks you've written feedback for in this group."
            transitionName="zoom-big-fast"
          >
            <Button
              block={false}
              ghost={false}
              htmlType="button"
              loading={false}
              type="link"
            >
              <Icon
                type="question-circle"
              />
            </Button>
          </Tooltip>
          <Progress
            format={[Function]}
            gapDegree={0}
            percent={75}
            showInfo={true}
            size="default"
            strokeLinecap="round"
            trailColor="#f3f3f3"
            type="line"
            width={80}
          />
        </Col>
      </Row>
      <Row
        gutter={
          Array [
            16,
            16,
          ]
        }
      >
        <Col>
          <Text
            strong={true}
          >
            Feedback for you
          </Text>
          <Tooltip
            arrowPointAtCenter={false}
            autoAdjustOverflow={true}
            mouseEnterDelay={0.1}
            mouseLeaveDelay={0.1}
            placement="top"
            title="How many members of this group have written feedback for your track (this number is hidden if you have not yet written feedback for everyone in your group)"
            transitionName="zoom-big-fast"
          >
            <Button
              block={false}
              ghost={false}
              htmlType="button"
              loading={false}
              type="link"
            >
              <Icon
                type="question-circle"
              />
            </Button>
          </Tooltip>
          <Progress
            format={[Function]}
            gapDegree={0}
            percent={60}
            showInfo={true}
            size="default"
            strokeLinecap="round"
            trailColor="#f3f3f3"
            type="line"
            width={80}
          />
        </Col>
      </Row>
      <Meta
        description={
          <CardExtra
            timeCreated={123456}
          />
        }
      />
    </A>
  </Card>
</Div>
`;
